packages/design-system/src/components/TextField/LabelMask/LabelMask.tsx

```diff
@@ -0,0 +1,84 @@
+import React from 'react';
+
+export enum BuiltInMask {
+  DATE = 'DATE',
+}
+
+export interface LabelMaskProps {
+  /**
+   * Must contain a `TextField` component
+   */
+  children: React.ReactNode;
+  labelMask?: BuiltInMask | ((rawInput: string) => string);
+}
+
+const LabelMask = ({ children, labelMask }: LabelMaskProps) => {
+  /**
+   * Get the child text field. Called as a method so that
+   * updates to the field cause the mask to re-render
+   * @returns {React.ReactElement} Child TextField
+   */
+  const field = (): React.ReactElement => React.Children.only(children as React.ReactElement);
+
+  /**
+   * @param {Object} evt
+   * @param {React.Element} field - Child TextField
+   */
+  const handleChange = (evt: React.ChangeEvent<HTMLInputElement>) => {
+    if (typeof field().props.onChange === 'function') {
+      return field().props.onChange(evt);
+    }
+  };
+
+  const modifiedTextField = React.cloneElement(field(), {
+    defaultValue: undefined,
+    onChange: (e) => handleChange(e),
+    type: 'text',
+    inputMode: 'numeric',
+  });
+
+  let hint = '';
+
+  function format(str): any {
+    if (labelMask === BuiltInMask.DATE) {
+      const charCount = str.length;
+
+      hint = 'MM/DD/YYYY';
+      const hintSub = hint.replace(/\//g, '').substring(charCount);
+
+      let val = str + hintSub;
+      // const dateRegex = /^(?<month>\d{1,2})(\s|-|\/)?(?<day>\d{0,2})(\s|-|\/)?(?<year>\d{0,4}).*/;
+      // const dateRegex = /^(?<month>\d{1,2})(?<day>\d{0,2})(?<year>\d{0,4}).*/;
+      const dateRegex = /^(?<month>[\S]{2})(?<day>[\S]{2})?(?<year>[\S]{1,4}).*/;
```

Comment:
I'm not crazy about this regex, but it was the only way I could allow the mixed number/letters format that happens while typing, ie, seeing "10/1D/YYY" in the hint text area. Without it, I'd get "10/1," which kinda defeats the purpose of a super intuitive mask while you type solution imo.

My regex-fu isn't very good though so I'm hopeful there's a better approach.

Reply:
You should be doing the regex on the input string before applying the mask

Reply:
Given `str` as your raw input (`props.value`):
1. Parse `str` for month, day, and year using one or more regular expression
    - There are two main scenarios: 1) they inputted delimiters or 2) they did not input delimiters
    - In both cases, something like `/^(\d{1,2})[\D]?(\d{1,2})?[-\D]?(\d{1,4})?/` should match, but in the second case we will not be able to detect if they omitted the leading zeros for the first two groups. That's just not something we can do 🤷 
    - Here's a sandbox using the regular expression above and a bunch of example input that I made up: https://regex101.com/r/9vRHLO/1
2. Once we have values month, day, and year (with the possibility that some are empty), we can format them into a string that matches our mask with something like `const date = `${month}/${day}/${year}`
3. We then mix our mask and our formatted input like you were doing before: `hint.replace(/\//g, '').substring(date.length);`

Reply:
This should get you your possibly-undefined variables for each of the groups
```js
const [month, day, year] = /^(\d{1,2})[\D]?(\d{1,2})?[-\D]?(\d{1,4})?/
      .exec(str)
      .slice(1)
      .filter(s => s)
```
and if you want them to default to empty strings, you can do
```js
const [month = '', day = '', year = ''] = ...
```

Reply:
I always have to look it up, but [there's a handy table for the returned array for RegExp's exec function here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/exec#description). The first element is the matched string, and we can ignore it (`.slice(1)`), and then we filter on only strings that exist (`.filter(s => s)`)...which I guess is not really necessary since we're destructuring the array...yeah, I guess you can remove that part; that was a straight copy from my phone number example in the ticket where I was iterating through the matched groups later.

Reply:
So more like
```
const [month = '', day = '', year = ''] = /^(\d{1,2})[\D]?(\d{1,2})?[-\D]?(\d{1,4})?/
      .exec(str)
      .slice(1);
```

Reply:
or
```js
const RE_DATE = /^(\d{1,2})[\D]?(\d{1,2})?[-\D]?(\d{1,4})?/;
// ...
const [month = '', day = '', year = ''] = RE_DATE.exec(str).slice(1);
```